### PR TITLE
Refactor `JSON::PullParser#consume_number` to use stdlib number parsing

### LINKED ISSUE #10419: `JSON::PullParser` is broken for `MIN` and `MAX` values
```cr
require "json"

pull = JSON::PullParser.new((Float64::MAX).to_json)
p pull.read?(Float64) # => Float64::INFINITY

pull = JSON::PullParser.new((Float32::MAX).to_json)
p pull.read?(Float32) # => nil
```

`read?` returns `nil` when an overflow error occurs. The overflow when casting `Float64::INFINITY` to `Float32` is a different bug (fixed in #10420).

Conversion is broken for MIN and MAX of both float types as well as `Int64::MIN`.

### LINKED ISSUE #10920: JSON parse gives wrong Float64 result
Repro steps
Below code show difference between from_json and to_f64. (result of to_f64 is consistent with other languages while from_json is not)
```crystal
num = "43.474709000000132"
puts Float64.from_json(num)
# 43.474709000000125
puts num.to_f64
# 43.47470900000013
```

Crystal compiler version
> Crystal 1.0.0 [dd40a2442] (2021-03-22)
LLVM: 10.0.0
Default target: x86_64-unknown-linux-gnu

### PR DESCRIPTION
**Pull request by straight-shoota**
The custom parsing algorithm was insufficient and had several issues around range boundaries.
Replacing it with the established conversion methods ensures correctness.

Resolves #10419
Resolves #10920

### CODE REVIEW

**Comment by asterite** [comment]
How is performance affected?

File: src/json/token.cr [hunk 1]
```diff
@@ -17,8 +17,23 @@ class JSON::Token
...
   property string_value : String
-  property int_value : Int64
-  property float_value : Float64
+
+  def int_value : Int64
+    begin
+      raw_value.to_i64
+    rescue exc : ArgumentError
+      raise ParseException.new(exc.message, line_number, column_number)
+    end
```
Comment (Sija) [inline src/json/token.cr:26]: Redundant `begin ... end`.

```suggestion
    raw_value.to_i64
  rescue exc : ArgumentError
    raise ParseException.new(exc.message, line_number, column_number)
```

File: src/json/token.cr [hunk 2]
```diff
@@ -17,8 +17,23 @@ class JSON::Token
...
+  end
+
+  def float_value : Float64
+    begin
+      raw_value.to_f64
+    rescue exc : ArgumentError
+      raise ParseException.new(exc.message, line_number, column_number)
+    end
```
Comment (Sija) [inline src/json/token.cr:34]: ditto

```suggestion
    raw_value.to_f64
  rescue exc : ArgumentError
    raise ParseException.new(exc.message, line_number, column_number)
```

**Comment by straight-shoota** [comment]
The focus is correctness, not performance. So I don't have an extensive performance analysis. It doesn't matter if we have a fast algorithm when its results are plain wrong. I'm sure this can be optimized further, but for now I'd just like to fix the bugs in number parsing.

These are the results for a simple test, parsing an array of 100 random ints/floats.
```
$ ./bm-json-number-old
  ints  73.67k ( 13.57µs) (± 3.63%)   5.5kB/op        fastest
floats  45.46k ( 22.00µs) (± 2.83%)  7.11kB/op   1.62× slower
$ ./bm-json-number-new
  ints  58.82k ( 17.00µs) (± 2.79%)  5.47kB/op        fastest
floats  31.69k ( 31.56µs) (± 2.92%)  7.09kB/op   1.86× slower
```

**Comment by kostya** [comment]
for this bench: https://github.com/kostya/benchmarks#json
was: 0.661 s
this pr: 0.839 s

**Comment by asterite** [comment]
@straight-shoota I believe correctness can be achieved without sacrificing performance. There's no need to go from a string to an intermediate string to an int or float. It can be parsed directly. Also numbers in JSON are simpler than what we cover in the std (no whitespace, no underscore, etc.)

**Comment by straight-shoota** [comment]
Sure, but this is not trivial. Until somebody implements that correctly, we should fall back to a less performant but correct implementation.

**Comment by kostya** [comment]
what if adapt this parser https://github.com/lemire/fast_double_parser, which is used by simdjson

**Comment by kostya** [comment]
for fun i try this, with this shard: https://github.com/kostya/fast_to_f

with such dirty patch:
```crystal
class JSON::Lexer::StringBased
  private def consume_number
    buf = @reader.string.to_unsafe + @reader.pos
    v, end_s = FastToF.parse_internal(buf)

    if end_s.null?
      unexpected_char
    else
      if v.round == v
        @token.kind = :int
        @token.int_value = v.to_i64
      else
        @token.kind = :float
        @token.float_value = v
      end

      chars_count = end_s - buf
      @reader.pos = @reader.pos + chars_count
    end
  end
end
```

my benchmark:
was: 0.655 s
with this: 0.550 s

**Review by sdogruyol — APPROVED** [review]
I'm on the same page with @straight-shoota here. I think the program should behave correct first, then we can always think about performance.

File: spec/std/json/pull_parser_spec.cr [hunk 3]
```diff
@@ -358,13 +366,67 @@ describe JSON::PullParser do
...
+      pull.read?(Int64).should be_nil
+    end
+
+    it "reads < Int64::MIN" do
+      pull = JSON::PullParser.new(Int64::MIN.to_s + "0")
+      pull.read?(Int64).should be_nil
+    end
+
+    it "reads > Float32::MAX" do
+      pull = JSON::PullParser.new(Float64::MAX.to_s)
```
Comment (drhuffman12) [inline spec/std/json/pull_parser_spec.cr:398]: Sorry if I'm missing something, but is this also supposed to have the ` + "0"` part?

Comment (straight-shoota) [reply spec/std/json/pull_parser_spec.cr:398]: I think you're missing that the tested value is `Float64::MAX`, not `Float32::MAX`.

**Comment by straight-shoota** [comment]
We'll postpone this to work on the implementation.

File: spec/std/json/pull_parser_spec.cr (hunk 3 above)
Comment (drhuffman12) [reply spec/std/json/pull_parser_spec.cr:398]: Sorry; I just now saw this; thanks. The appended `0` for the `reads > Int64::MAX` test makes sense now; that would make the `1.7976931348623157e+308` value into `1.7976931348623157e+3080`, which would be more positive. Likewise, for the `reads < Int64::MIN` test, that would make the `-1.7976931348623157e+308` value into `-1.7976931348623157e+3080`, which would be more negative.

**Comment by straight-shoota** [comment]
After reviewing this again, I think we should merge this as is. This patch fixes a number of serious bugs in JSON number parsing. Performance degradation is insignificant over ensuring correctness.
Sure, this can be implemented more efficiently and we should aim to do that. But that's gonna take some more effort. We must not postpone the low hanging fruit of fixing the parser while waiting for a more performant implementation.

**Review by beta-ziliani — APPROVED** [review]
Given that there's no better proposal yet, I would like this one to get merged. Like @straight-shoota said, better a correct algorithm than a broken one. When we get a new working _and_ fast algorithm, we'll replace this one.

**Pull request merged**